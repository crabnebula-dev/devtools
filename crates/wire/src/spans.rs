use crate::common;

mod generated {
    #![allow(warnings)]
    #![allow(clippy::all, clippy::pedantic)]
    include!("./generated/rs.devtools.spans.rs");
}

pub use generated::*;

impl SpanEvent {
    #[must_use]
    pub fn new_span(
        at: prost_types::Timestamp,
        id: &tracing_core::span::Id,
        metadata: &'static tracing_core::Metadata<'static>,
        fields: Vec<common::Field>,
        parent: Option<tracing_core::span::Id>,
    ) -> Self {
        Self {
            event: Some(span_event::Event::NewSpan(span_event::Span {
                id: id.into_u64(),
                metadata_id: std::ptr::from_ref(metadata) as u64,
                fields,
                at: Some(at),
                parent: parent.map(|id| id.into_u64()),
            })),
        }
    }

    #[must_use]
    pub fn enter_span(
        at: prost_types::Timestamp,
        id: &tracing_core::span::Id,
        thread_id: u64,
    ) -> Self {
        Self {
            event: Some(span_event::Event::EnterSpan(span_event::Enter {
                span_id: id.into_u64(),
                thread_id,
                at: Some(at),
            })),
        }
    }

    #[must_use]
    pub fn exit_span(
        at: prost_types::Timestamp,
        id: &tracing_core::span::Id,
        thread_id: u64,
    ) -> Self {
        Self {
            event: Some(span_event::Event::ExitSpan(span_event::Exit {
                span_id: id.into_u64(),
                thread_id,
                at: Some(at),
            })),
        }
    }

    #[must_use]
    pub fn close_span(at: prost_types::Timestamp, id: &tracing_core::span::Id) -> Self {
        Self {
            event: Some(span_event::Event::CloseSpan(span_event::Close {
                span_id: id.into_u64(),
                at: Some(at),
            })),
        }
    }

    #[must_use]
    pub fn span_recorded(id: &tracing_core::span::Id, fields: Vec<common::Field>) -> Self {
        Self {
            event: Some(span_event::Event::Recorded(span_event::Recorded {
                span_id: id.into_u64(),
                fields,
            })),
        }
    }
}
