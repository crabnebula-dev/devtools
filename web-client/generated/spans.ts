// @generated by protobuf-ts 2.9.1
// @generated from protobuf file "spans.proto" (package "rs.devtools.spans", syntax proto3)
// tslint:disable
import type { BinaryWriteOptions } from "@protobuf-ts/runtime";
import type { IBinaryWriter } from "@protobuf-ts/runtime";
import { WireType } from "@protobuf-ts/runtime";
import type { BinaryReadOptions } from "@protobuf-ts/runtime";
import type { IBinaryReader } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import type { PartialMessage } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MESSAGE_TYPE } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Timestamp } from "./google/protobuf/timestamp";
import { Field } from "./common";
import { MetaId } from "./common";
import { SpanId } from "./common";
/**
 * @generated from protobuf message rs.devtools.spans.Update
 */
export interface Update {
    /**
     * @generated from protobuf field: repeated rs.devtools.spans.SpanEvent span_events = 1;
     */
    spanEvents: SpanEvent[];
    /**
     * @generated from protobuf field: uint64 dropped_events = 2;
     */
    droppedEvents: bigint;
}
/**
 * @generated from protobuf message rs.devtools.spans.SpanEvent
 */
export interface SpanEvent {
    /**
     * @generated from protobuf oneof: event
     */
    event: {
        oneofKind: "newSpan";
        /**
         * @generated from protobuf field: rs.devtools.spans.SpanEvent.Span new_span = 1;
         */
        newSpan: SpanEvent_Span;
    } | {
        oneofKind: "enterSpan";
        /**
         * @generated from protobuf field: rs.devtools.spans.SpanEvent.Enter enter_span = 2;
         */
        enterSpan: SpanEvent_Enter;
    } | {
        oneofKind: "exitSpan";
        /**
         * @generated from protobuf field: rs.devtools.spans.SpanEvent.Exit exit_span = 3;
         */
        exitSpan: SpanEvent_Exit;
    } | {
        oneofKind: "closeSpan";
        /**
         * @generated from protobuf field: rs.devtools.spans.SpanEvent.Close close_span = 4;
         */
        closeSpan: SpanEvent_Close;
    } | {
        oneofKind: undefined;
    };
}
/**
 * Represents a period of time in which a program was executing in a particular context.
 *
 * Corresponds to `Span` in the `tracing` crate.
 *
 * @generated from protobuf message rs.devtools.spans.SpanEvent.Span
 */
export interface SpanEvent_Span {
    /**
     * An Id that uniquely identifies it in relation to other spans.
     *
     * @generated from protobuf field: rs.devtools.common.SpanId id = 1;
     */
    id?: SpanId;
    /**
     * Identifier for metadata describing static characteristics of all spans originating
     * from that callsite, such as its name, source code location, verbosity level, and
     * the names of its fields.
     *
     * @generated from protobuf field: rs.devtools.common.MetaId metadata_id = 2;
     */
    metadataId?: MetaId;
    /**
     * User-defined key-value pairs of arbitrary data that describe the context the span represents,
     *
     * @generated from protobuf field: repeated rs.devtools.common.Field fields = 3;
     */
    fields: Field[];
    /**
     * @generated from protobuf field: optional rs.devtools.common.SpanId parent = 4;
     */
    parent?: SpanId;
    /**
     * Timestamp for the span.
     *
     * @generated from protobuf field: google.protobuf.Timestamp at = 5;
     */
    at?: Timestamp;
}
/**
 * @generated from protobuf message rs.devtools.spans.SpanEvent.Enter
 */
export interface SpanEvent_Enter {
    /**
     * @generated from protobuf field: rs.devtools.common.SpanId span_id = 1;
     */
    spanId?: SpanId;
    /**
     * @generated from protobuf field: uint64 thread_id = 2;
     */
    threadId: bigint;
    /**
     * @generated from protobuf field: google.protobuf.Timestamp at = 3;
     */
    at?: Timestamp;
}
/**
 * @generated from protobuf message rs.devtools.spans.SpanEvent.Exit
 */
export interface SpanEvent_Exit {
    /**
     * @generated from protobuf field: rs.devtools.common.SpanId span_id = 1;
     */
    spanId?: SpanId;
    /**
     * @generated from protobuf field: uint64 thread_id = 2;
     */
    threadId: bigint;
    /**
     * @generated from protobuf field: google.protobuf.Timestamp at = 3;
     */
    at?: Timestamp;
}
/**
 * @generated from protobuf message rs.devtools.spans.SpanEvent.Close
 */
export interface SpanEvent_Close {
    /**
     * @generated from protobuf field: rs.devtools.common.SpanId span_id = 1;
     */
    spanId?: SpanId;
    /**
     * @generated from protobuf field: google.protobuf.Timestamp at = 3;
     */
    at?: Timestamp;
}
// @generated message type with reflection information, may provide speed optimized methods
class Update$Type extends MessageType<Update> {
    constructor() {
        super("rs.devtools.spans.Update", [
            { no: 1, name: "span_events", kind: "message", repeat: 1 /*RepeatType.PACKED*/, T: () => SpanEvent },
            { no: 2, name: "dropped_events", kind: "scalar", T: 4 /*ScalarType.UINT64*/, L: 0 /*LongType.BIGINT*/ }
        ]);
    }
    create(value?: PartialMessage<Update>): Update {
        const message = { spanEvents: [], droppedEvents: 0n };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial<Update>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: Update): Update {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* repeated rs.devtools.spans.SpanEvent span_events */ 1:
                    message.spanEvents.push(SpanEvent.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                case /* uint64 dropped_events */ 2:
                    message.droppedEvents = reader.uint64().toBigInt();
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: Update, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* repeated rs.devtools.spans.SpanEvent span_events = 1; */
        for (let i = 0; i < message.spanEvents.length; i++)
            SpanEvent.internalBinaryWrite(message.spanEvents[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* uint64 dropped_events = 2; */
        if (message.droppedEvents !== 0n)
            writer.tag(2, WireType.Varint).uint64(message.droppedEvents);
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message rs.devtools.spans.Update
 */
export const Update = new Update$Type();
// @generated message type with reflection information, may provide speed optimized methods
class SpanEvent$Type extends MessageType<SpanEvent> {
    constructor() {
        super("rs.devtools.spans.SpanEvent", [
            { no: 1, name: "new_span", kind: "message", oneof: "event", T: () => SpanEvent_Span },
            { no: 2, name: "enter_span", kind: "message", oneof: "event", T: () => SpanEvent_Enter },
            { no: 3, name: "exit_span", kind: "message", oneof: "event", T: () => SpanEvent_Exit },
            { no: 4, name: "close_span", kind: "message", oneof: "event", T: () => SpanEvent_Close }
        ]);
    }
    create(value?: PartialMessage<SpanEvent>): SpanEvent {
        const message = { event: { oneofKind: undefined } };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial<SpanEvent>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: SpanEvent): SpanEvent {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* rs.devtools.spans.SpanEvent.Span new_span */ 1:
                    message.event = {
                        oneofKind: "newSpan",
                        newSpan: SpanEvent_Span.internalBinaryRead(reader, reader.uint32(), options, (message.event as any).newSpan)
                    };
                    break;
                case /* rs.devtools.spans.SpanEvent.Enter enter_span */ 2:
                    message.event = {
                        oneofKind: "enterSpan",
                        enterSpan: SpanEvent_Enter.internalBinaryRead(reader, reader.uint32(), options, (message.event as any).enterSpan)
                    };
                    break;
                case /* rs.devtools.spans.SpanEvent.Exit exit_span */ 3:
                    message.event = {
                        oneofKind: "exitSpan",
                        exitSpan: SpanEvent_Exit.internalBinaryRead(reader, reader.uint32(), options, (message.event as any).exitSpan)
                    };
                    break;
                case /* rs.devtools.spans.SpanEvent.Close close_span */ 4:
                    message.event = {
                        oneofKind: "closeSpan",
                        closeSpan: SpanEvent_Close.internalBinaryRead(reader, reader.uint32(), options, (message.event as any).closeSpan)
                    };
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: SpanEvent, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* rs.devtools.spans.SpanEvent.Span new_span = 1; */
        if (message.event.oneofKind === "newSpan")
            SpanEvent_Span.internalBinaryWrite(message.event.newSpan, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* rs.devtools.spans.SpanEvent.Enter enter_span = 2; */
        if (message.event.oneofKind === "enterSpan")
            SpanEvent_Enter.internalBinaryWrite(message.event.enterSpan, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        /* rs.devtools.spans.SpanEvent.Exit exit_span = 3; */
        if (message.event.oneofKind === "exitSpan")
            SpanEvent_Exit.internalBinaryWrite(message.event.exitSpan, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
        /* rs.devtools.spans.SpanEvent.Close close_span = 4; */
        if (message.event.oneofKind === "closeSpan")
            SpanEvent_Close.internalBinaryWrite(message.event.closeSpan, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message rs.devtools.spans.SpanEvent
 */
export const SpanEvent = new SpanEvent$Type();
// @generated message type with reflection information, may provide speed optimized methods
class SpanEvent_Span$Type extends MessageType<SpanEvent_Span> {
    constructor() {
        super("rs.devtools.spans.SpanEvent.Span", [
            { no: 1, name: "id", kind: "message", T: () => SpanId },
            { no: 2, name: "metadata_id", kind: "message", T: () => MetaId },
            { no: 3, name: "fields", kind: "message", repeat: 1 /*RepeatType.PACKED*/, T: () => Field },
            { no: 4, name: "parent", kind: "message", T: () => SpanId },
            { no: 5, name: "at", kind: "message", T: () => Timestamp }
        ]);
    }
    create(value?: PartialMessage<SpanEvent_Span>): SpanEvent_Span {
        const message = { fields: [] };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial<SpanEvent_Span>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: SpanEvent_Span): SpanEvent_Span {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* rs.devtools.common.SpanId id */ 1:
                    message.id = SpanId.internalBinaryRead(reader, reader.uint32(), options, message.id);
                    break;
                case /* rs.devtools.common.MetaId metadata_id */ 2:
                    message.metadataId = MetaId.internalBinaryRead(reader, reader.uint32(), options, message.metadataId);
                    break;
                case /* repeated rs.devtools.common.Field fields */ 3:
                    message.fields.push(Field.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                case /* optional rs.devtools.common.SpanId parent */ 4:
                    message.parent = SpanId.internalBinaryRead(reader, reader.uint32(), options, message.parent);
                    break;
                case /* google.protobuf.Timestamp at */ 5:
                    message.at = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.at);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: SpanEvent_Span, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* rs.devtools.common.SpanId id = 1; */
        if (message.id)
            SpanId.internalBinaryWrite(message.id, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* rs.devtools.common.MetaId metadata_id = 2; */
        if (message.metadataId)
            MetaId.internalBinaryWrite(message.metadataId, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        /* repeated rs.devtools.common.Field fields = 3; */
        for (let i = 0; i < message.fields.length; i++)
            Field.internalBinaryWrite(message.fields[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
        /* optional rs.devtools.common.SpanId parent = 4; */
        if (message.parent)
            SpanId.internalBinaryWrite(message.parent, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
        /* google.protobuf.Timestamp at = 5; */
        if (message.at)
            Timestamp.internalBinaryWrite(message.at, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message rs.devtools.spans.SpanEvent.Span
 */
export const SpanEvent_Span = new SpanEvent_Span$Type();
// @generated message type with reflection information, may provide speed optimized methods
class SpanEvent_Enter$Type extends MessageType<SpanEvent_Enter> {
    constructor() {
        super("rs.devtools.spans.SpanEvent.Enter", [
            { no: 1, name: "span_id", kind: "message", T: () => SpanId },
            { no: 2, name: "thread_id", kind: "scalar", T: 4 /*ScalarType.UINT64*/, L: 0 /*LongType.BIGINT*/ },
            { no: 3, name: "at", kind: "message", T: () => Timestamp }
        ]);
    }
    create(value?: PartialMessage<SpanEvent_Enter>): SpanEvent_Enter {
        const message = { threadId: 0n };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial<SpanEvent_Enter>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: SpanEvent_Enter): SpanEvent_Enter {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* rs.devtools.common.SpanId span_id */ 1:
                    message.spanId = SpanId.internalBinaryRead(reader, reader.uint32(), options, message.spanId);
                    break;
                case /* uint64 thread_id */ 2:
                    message.threadId = reader.uint64().toBigInt();
                    break;
                case /* google.protobuf.Timestamp at */ 3:
                    message.at = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.at);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: SpanEvent_Enter, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* rs.devtools.common.SpanId span_id = 1; */
        if (message.spanId)
            SpanId.internalBinaryWrite(message.spanId, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* uint64 thread_id = 2; */
        if (message.threadId !== 0n)
            writer.tag(2, WireType.Varint).uint64(message.threadId);
        /* google.protobuf.Timestamp at = 3; */
        if (message.at)
            Timestamp.internalBinaryWrite(message.at, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message rs.devtools.spans.SpanEvent.Enter
 */
export const SpanEvent_Enter = new SpanEvent_Enter$Type();
// @generated message type with reflection information, may provide speed optimized methods
class SpanEvent_Exit$Type extends MessageType<SpanEvent_Exit> {
    constructor() {
        super("rs.devtools.spans.SpanEvent.Exit", [
            { no: 1, name: "span_id", kind: "message", T: () => SpanId },
            { no: 2, name: "thread_id", kind: "scalar", T: 4 /*ScalarType.UINT64*/, L: 0 /*LongType.BIGINT*/ },
            { no: 3, name: "at", kind: "message", T: () => Timestamp }
        ]);
    }
    create(value?: PartialMessage<SpanEvent_Exit>): SpanEvent_Exit {
        const message = { threadId: 0n };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial<SpanEvent_Exit>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: SpanEvent_Exit): SpanEvent_Exit {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* rs.devtools.common.SpanId span_id */ 1:
                    message.spanId = SpanId.internalBinaryRead(reader, reader.uint32(), options, message.spanId);
                    break;
                case /* uint64 thread_id */ 2:
                    message.threadId = reader.uint64().toBigInt();
                    break;
                case /* google.protobuf.Timestamp at */ 3:
                    message.at = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.at);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: SpanEvent_Exit, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* rs.devtools.common.SpanId span_id = 1; */
        if (message.spanId)
            SpanId.internalBinaryWrite(message.spanId, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* uint64 thread_id = 2; */
        if (message.threadId !== 0n)
            writer.tag(2, WireType.Varint).uint64(message.threadId);
        /* google.protobuf.Timestamp at = 3; */
        if (message.at)
            Timestamp.internalBinaryWrite(message.at, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message rs.devtools.spans.SpanEvent.Exit
 */
export const SpanEvent_Exit = new SpanEvent_Exit$Type();
// @generated message type with reflection information, may provide speed optimized methods
class SpanEvent_Close$Type extends MessageType<SpanEvent_Close> {
    constructor() {
        super("rs.devtools.spans.SpanEvent.Close", [
            { no: 1, name: "span_id", kind: "message", T: () => SpanId },
            { no: 3, name: "at", kind: "message", T: () => Timestamp }
        ]);
    }
    create(value?: PartialMessage<SpanEvent_Close>): SpanEvent_Close {
        const message = {};
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial<SpanEvent_Close>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: SpanEvent_Close): SpanEvent_Close {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* rs.devtools.common.SpanId span_id */ 1:
                    message.spanId = SpanId.internalBinaryRead(reader, reader.uint32(), options, message.spanId);
                    break;
                case /* google.protobuf.Timestamp at */ 3:
                    message.at = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.at);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: SpanEvent_Close, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* rs.devtools.common.SpanId span_id = 1; */
        if (message.spanId)
            SpanId.internalBinaryWrite(message.spanId, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* google.protobuf.Timestamp at = 3; */
        if (message.at)
            Timestamp.internalBinaryWrite(message.at, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message rs.devtools.spans.SpanEvent.Close
 */
export const SpanEvent_Close = new SpanEvent_Close$Type();
