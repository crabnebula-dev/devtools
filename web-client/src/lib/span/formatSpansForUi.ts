import { Span } from "../connection/monitor";
import { Metadata } from "../proto/common";
import { getIpcRequestName } from "./getIpcRequestName";
import { normalizeSpans } from "./normalizeSpans";

type Options = {
    spans: Span[]
    metadata: Map<bigint, Metadata>
}
export function formatSpansForUi({ spans, metadata }: Options) {
    return normalizeSpans(spans).map(span => {

        return ({
            name: getIpcRequestName({ metadata, span }) || '-',
            initiated: span.createdAt / 1000000,
            time: span.closedAt ? (span.closedAt - span.createdAt) / 1e6 : 0,
            waterfall: `width:${span.width}%;margin-left:${span.marginLeft}%;`,
            start: span.marginLeft,
            slices: span.slices.map(slice => `width:${slice.width}%;margin-left:${slice.marginLeft}%;`)
        })
    })
}