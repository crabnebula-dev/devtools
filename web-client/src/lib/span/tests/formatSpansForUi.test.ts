import { describe, expect } from "vitest";
import { formatSpansForUi } from "../formatSpansForUi";
import { spans } from "./fixtures/spans";
import { metadata } from "./fixtures/metadata";

describe("formatSpansForUi", () => {
  it("should format spans for UI", () => {
    // @ts-expect-error this is a test...
    expect(formatSpansForUi({ spans, metadata })).toMatchSnapshot();
  });
});
