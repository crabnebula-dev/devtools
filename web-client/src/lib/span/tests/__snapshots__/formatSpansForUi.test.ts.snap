// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatSpansForUi > should format spans for UI 1`] = `
[
  {
    "children": [],
    "colorClassName": "bg-red-500",
    "id": "11",
    "initiated": 1700085057720.6262,
    "name": "-",
    "slices": [
      "width:100%;margin-left:-5.421010862427522e-20%;",
    ],
    "start": 0,
    "time": 2337.679104,
    "waterfall": "width:100%;margin-left:0%;",
  },
  {
    "children": [],
    "colorClassName": "bg-red-500",
    "id": "12",
    "initiated": 1700085057720.6707,
    "name": "-",
    "slices": [
      "width:100%;margin-left:0%;",
    ],
    "start": 0.001905479666725036,
    "time": 2337.632512,
    "waterfall": "width:99.99800691207274%;margin-left:0.001905479666725036%;",
  },
  {
    "children": [],
    "colorClassName": "bg-red-500",
    "id": "13",
    "initiated": 1700085057720.7607,
    "name": "-",
    "slices": [
      "width:100%;margin-left:0%;",
    ],
    "start": 0.005760243130444648,
    "time": 2337.540352,
    "waterfall": "width:99.99406454034848%;margin-left:0.005760243130444648%;",
  },
  {
    "children": [],
    "colorClassName": "bg-red-500",
    "id": "14",
    "initiated": 1700085057720.7712,
    "name": "-",
    "slices": [
      "width:0.8600315126050421%;margin-left:0.004297766708984177%;",
      "width:9.158350840336134%;margin-left:77.95807843925697%;",
      "width:3.9915966386554618%;margin-left:82.49167468061351%;",
      "width:3.3350840336134455%;margin-left:84.08707116870455%;",
      "width:1.3950892857142858%;margin-left:86.9385981550531%;",
      "width:70.01050420168067%;margin-left:26.46460007638799%;",
      "width:4.237788865546219%;margin-left:86.719700909331%;",
      "width:1.4935661764705883%;margin-left:89.22199581787801%;",
      "width:3.12828256302521%;margin-left:92.99287298905453%;",
      "width:0.7910976890756303%;margin-left:95.25068337027949%;",
      "width:0.2658876050420168%;margin-left:99.72767636557269%;",
      "width:1.2867647058823528%;margin-left:98.70844684782573%;",
      "width:0.06236869747899159%;margin-left:99.93728032767245%;",
    ],
    "start": 0.0062092354657074435,
    "time": 2332.614656,
    "waterfall": "width:99.78335572271942%;margin-left:0.0062092354657074435%;",
  },
  {
    "children": [],
    "colorClassName": "bg-teal-500",
    "id": "4504149383184387",
    "initiated": 1700085057720.8862,
    "name": "-",
    "slices": [
      "width:100%;margin-left:-7.105427357601002e-15%;",
    ],
    "start": 0.011126249088463427,
    "time": 0.01536,
    "waterfall": "width:0.0006570619540431157%;margin-left:0.011126249088463427%;",
  },
  {
    "children": [],
    "colorClassName": "bg-teal-500",
    "id": "6755949196869635",
    "initiated": 1700085057720.911,
    "name": "-",
    "slices": [
      "width:100%;margin-left:0%;",
    ],
    "start": 0.012177548214932412,
    "time": 0.005376,
    "waterfall": "width:0.00022997168391509053%;margin-left:0.012177548214932412%;",
  },
  {
    "children": [],
    "colorClassName": "bg-red-500",
    "id": "15",
    "initiated": 1700085057720.775,
    "name": "-",
    "slices": [
      "width:0.5941460973982655%;margin-left:0.003951868437680539%;",
      "width:5.947715143428953%;margin-left:80.54327479999732%;",
      "width:2.770597064709807%;margin-left:83.36487232834703%;",
      "width:2.2473315543695795%;margin-left:84.85411015224128%;",
      "width:1.0361074049366243%;margin-left:87.07121327426503%;",
      "width:44.57346564376251%;margin-left:48.80887195638774%;",
      "width:2.8122915276851232%;margin-left:87.82516631046421%;",
      "width:1.025683789192795%;margin-left:89.45721088482225%;",
      "width:2.120163442294863%;margin-left:93.76281642767363%;",
      "width:0.6212474983322215%;margin-left:95.21294266861659%;",
      "width:0.2501667778519013%;margin-left:99.53354178488352%;",
      "width:35.969813208805874%;margin-left:63.89242055258796%;",
      "width:0.033355570380253496%;margin-left:99.94069741759392%;",
    ],
    "start": 0.006362549921650837,
    "time": 2337.523968,
    "waterfall": "width:99.99336367426417%;margin-left:0.006362549921650837%;",
  },
  {
    "children": [],
    "colorClassName": "bg-teal-500",
    "id": "9007749010554884",
    "initiated": 1700085060054.4126,
    "name": "-",
    "slices": [
      "width:100%;margin-left:0%;",
    ],
    "start": 99.83347859878035,
    "time": 3.884288,
    "waterfall": "width:0.16616001714493658%;margin-left:99.83347859878035%;",
  },
  {
    "children": [],
    "colorClassName": "bg-teal-500",
    "id": "549755813892",
    "initiated": 1700085060054.459,
    "name": "-",
    "slices": [
      "width:88.99514701684271%;margin-left:0.018364683571118923%;",
      "width:11.004852983157294%;margin-left:65.33993872493511%;",
    ],
    "start": 99.83546073567504,
    "time": 3.835136,
    "waterfall": "width:0.1640574188919986%;margin-left:99.83546073567504%;",
  },
  {
    "children": [],
    "colorClassName": "bg-teal-500",
    "id": "16",
    "initiated": 1700085060057.302,
    "name": "-",
    "slices": [
      "width:100%;margin-left:0%;",
    ],
    "start": 99.95708290336842,
    "time": 0.987136,
    "waterfall": "width:0.042227181579837576%;margin-left:99.95708290336842%;",
  },
]
`;
